plugins/pulp_rpm/plugins/db/models.py

```diff
@@ -188,6 +193,44 @@ def __cmp__(self, other):
             other.complete_version_serialized
         )
 
+    def get_or_calculate_checksum(self, checksumtype):
+        """
+        Tries to get the sha1 or sha256 checksum value from the attribute on self of the
+        corresponding name. If it is not available, calculates the value and saves it on self
+        before returning it.
+
+        SIDE EFFECT: This method will save the model if the requested checksumtype is not populated
+                     on the model in the "checksums" attribute.
+
+        :param checksumtype:   any checksum type supported by the platform
+        :type  checksumtype:   basestring
+
+        :return:    checksum value of the requested type
+        :rtype:     basestring
+
+        :raises PulpCodedException: if the checksumtype is not available
+        """
+        if checksumtype not in server_util.CHECKSUM_FUNCTIONS:
+            raise ValueError(_('Checksum type %(checksumtype)s is not supported'))
```

Comment:
I could see it both ways. ``PulpCodedException`` should be used to give bad news to a user. In theory, any time a user specifies a checksum type, that should be validated closer to the edges.

For publishes in this plugin, the config gets validated, which includes the checksum type. So an unsupported checksum type getting to this level is most likely to be a programming error, not a user input error.

That said, there may not be any harm in raising a PulpCodedException. Perhaps the only risk is that a user might see an error message suggesting they did something wrong, when really they didn't. Although that's probably not worse than seeing an unhandled ValueError.

It's difficult to know where to draw the line. I think we could probably talk ourselves into only ever raising ``PulpCodedExeption``s, but then that would diminish their value.

Any of that making sense? What do you think?